storage: Added txn wait queue related metrics

darinpp opened the pull request:
Resolves https://github.com/cockroachdb/cockroach/issues/31106

The new metrics are:
- current count of pushees, pushers and queries
- the total number of instances where a pusher had to wait more than 1 min on the queue
- histograms for the time spent waiting on the queue for pushers and queries
- total number of deadlocks

Here is a sample chart for the current count of pushees, pushers, queries as well as the total number of cases where a pusher waited more than 1 min
![image](https://user-images.githubusercontent.com/7052693/54570571-3c7c0f80-499c-11e9-938e-e58cda791e17.png)

Here is a sample chart of the cumulative count of the deadlocks expressed as rate
![image](https://user-images.githubusercontent.com/7052693/54570649-895fe600-499c-11e9-9e56-c24a3dad52b6.png)

Here is a sample chart of the pusher wait time breakdown 
![image](https://user-images.githubusercontent.com/7052693/54570704-bb714800-499c-11e9-9a8f-7da56f697997.png)

and a sample chart of the queue wait time breakdown
![image](https://user-images.githubusercontent.com/7052693/54570731-d643bc80-499c-11e9-8fc5-3ed6f39344ea.png)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35921)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=35921) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=35921) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=35921) it.</sub>

darinpp commented:
I haven't added any tests. Not sure how to test these. Suggestions are welcome.

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txnwait/txnqueue.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaIHZnY8QrvTLc0lM-V:-LaIHZnY8QrvTLc0lM-W:b-gg00wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L473)):*
> ```Go
> 
> 	excessiveWait := time.Minute
> 	excessiveWaitTimer := time.NewTimer(excessiveWait)
> ```

Is there any concern that we're spinning up a goroutine each time we get to this point in the code? Is this on a critical perf path?


<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
> I haven't added any tests. Not sure how to test these. Suggestions are welcome.

I would start by instrumenting the existing tests so that they verify the metrics at each step in the test. Additionally, there are certain invariants that should be true at the beginning end of all tests when a queue is empty. (i.e. all gauges should be zero, etc).

BTW, the existing tests are in `pkg/storage/txn_wait_queue_test.go` for historical reasons (tech debt).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-:-LaLFWud3Zyy3C6jpBrA:b-hfvu0r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txnwait/metrics.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLAoLpChwIYnOJnVYr:-LaLAoLq4237mUozYCmf:b-58vwea) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

2019 (the current year is used for new code)
___
*[pkg/storage/txnwait/metrics.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLBC9NCX8b4zgI-7ri:-LaLBC9NCX8b4zgI-7rj:boogzv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L37)):*
> ```Go
> 		Pushees: metric.NewGauge(
> 			metric.Metadata{
> 				Name:        "txn_wait_queue.pushees",
> ```

We're wildly inconsistent in our naming of metrics, but I'd say

1. `txnwaitqueue` throughout
2. `txnwaitqueue.{pushee,pusher,query}.waiting`
___
*[pkg/storage/txnwait/metrics.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLBkxV-t3seEmUZpX2:-LaLBkxV-t3seEmUZpX3:b-g04boy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L64)):*
> ```Go
> 		PusherExcessiveWaitTotal: metric.NewCounter(
> 			metric.Metadata{
> 				Name:        "txn_wait_queue.pusher.excessive_wait_total",
> ```

Taking

```
root@127.0.0.1:54192/defaultdb> select name from crdb_internal.node_metrics WHERE name LIKE '%slow%';
         name
+---------------------+
  requests.slow.lease
  requests.slow.latch
  requests.slow.raft
(3 rows)
```

as a precedent, make this `txnwaitqueue.pusher.slow`

Also, this should be a gauge. In normal operation, it's always zero. It's nonzero exactly while a request is slow. (This is how all of the other ones work).
___
*[pkg/storage/txnwait/metrics.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLC_HB4J1Or30mbkTW:-LaLC_HC91qJHTtU9X4G:btn2zru) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L74)):*
> Histogram of durations spent in queue by pushers.
___
*[pkg/storage/txnwait/metrics.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLCuGfAdrzn1EWxQZ4:-LaLCuGfAdrzn1EWxQZ5:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L86)):*
> ```Go
> 			metric.Metadata{
> 				Name:        "txn_wait_queue.query.wait_time",
> 				Help:        "The breakdown of the query's wait time when they are removed from the queue",
> ```

Ditto.
___
*[pkg/storage/txnwait/metrics.go, line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLCwG907Qf5mbRKNDg:-LaLCwG907Qf5mbRKNDh:bp4arkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L98)):*
> Number of deadlocks detected by the txn wait queue.
___
*[pkg/storage/txnwait/txnqueue.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaIHZnY8QrvTLc0lM-V:-LaLDQ-REDmfELKvIqZV:b-tui5po) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L473)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Is there any concern that we're spinning up a goroutine each time we get to this point in the code? Is this on a critical perf path?
</blockquote></details>

Please use the same pattern from

https://github.com/cockroachdb/cockroach/blob/38e796199562090d2df956ad643e71b25ee25915/pkg/storage/replica_write.go#L194-L223

which is allocation optimized (and prevents deadlocks).

@andy-kimball if you enter the queue, a lot more work is already being done than this txn, so it shouldn't be on the critical perf path.
___
*[pkg/storage/txnwait/txnqueue.go, line 481 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLEpRf6d5fuL2j-wzq:-LaLEpRf6d5fuL2j-wzr:b6oz9xi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L481)):*
> ```Go
> 	// queue.
> 	pusheeTxnTimer.Reset(0)
> 	startWaiting := time.Now()
> ```

Instead of calling `RecordValue` for this in many different places, `defer` a function here that does it.
___
*[pkg/storage/txnwait/txnqueue.go, line 485 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLDm5-0gE8-9dgJhjD:-LaLDm5-0gE8-9dgJhjE:bst3mom) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L485)):*
> ```Go
> 		select {
> 		case <- excessiveWaitTimer.C:
> 			log.Warningf(ctx, "pusher request longer than %s", excessiveWait)
> ```

Please also use the same `have been waiting` kind of wording and print the pusher and pushee. You also need to `defer` a decrement to the gauge. Basically just do exactly what that other code I refer to above does.
___
*[pkg/storage/txnwait/txnqueue.go, line 694 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLFAucENMsOvlH1f4O:-LaLFAucENMsOvlH1f4P:bsgvzgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L694)):*
> ```Go
> 	select {
> 	case <-ctx.Done():
> 		metrics.QueryWaitTime.RecordValue(time.Since(startWaiting).Nanoseconds())
> ```

Ditto about `defer`ing this once instead.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> Reviewed 2 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, and @nvanbenschoten)

___
*[pkg/storage/txnwait/metrics.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaMdfb9-hUXMGIz_mOh:-LaMdfb9-hUXMGIz_mOi:blvy4lt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L56)):*
> ```Go
> 			metric.Metadata{
> 				Name:        "txn_wait_queue.queries",
> 				Help:        "Number of queries on the txn wait queue",
> ```

Can we say more about what a "query" is here? "Pushee" and "pusher" make sense but without going back to the code I can't remember what a "query" is. 


<!-- Sent from Reviewable.io -->

darinpp committed:
storage: Added txn wait queue related metrics

Resolves #31106

The new metrics are:

- current count of pushees, pushers and queries
- the total number of instances where a pusher had to wait more than 1 min on the queue
- histograms for the time spent waiting on the queue for pushers and queries
- total number of deadlocks

Release note (admin ui change): Added txn wait queue related metrics for debugging (available only via the custom chart page)

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txnwait/metrics.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLAoLpChwIYnOJnVYr:-LaYnTa50kk7dTMJZvZL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L1)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

2019 (the current year is used for new code)
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/metrics.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLBC9NCX8b4zgI-7ri:-LaYnaZz48mhlvF59z-M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L37)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We're wildly inconsistent in our naming of metrics, but I'd say

1. `txnwaitqueue` throughout
2. `txnwaitqueue.{pushee,pusher,query}.waiting`
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/metrics.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaMdfb9-hUXMGIz_mOh:-LaYnbi71C-T5AwFNAew:b-fs5ejn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L56)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can we say more about what a "query" is here? "Pushee" and "pusher" make sense but without going back to the code I can't remember what a "query" is. 
</blockquote></details>

added some clarification based on my understanding.
___
*[pkg/storage/txnwait/metrics.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLBkxV-t3seEmUZpX2:-LaYnokq-TDZtTCLrW8_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L64)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Taking

```
root@127.0.0.1:54192/defaultdb> select name from crdb_internal.node_metrics WHERE name LIKE '%slow%';
         name
+---------------------+
  requests.slow.lease
  requests.slow.latch
  requests.slow.raft
(3 rows)
```

as a precedent, make this `txnwaitqueue.pusher.slow`

Also, this should be a gauge. In normal operation, it's always zero. It's nonzero exactly while a request is slow. (This is how all of the other ones work).
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/metrics.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLC_HB4J1Or30mbkTW:-LaYnpnj2EVqT1kBF9PO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L74)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> Histogram of durations spent in queue by pushers.
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/metrics.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLCuGfAdrzn1EWxQZ4:-LaYnrstBba0memHJdov:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L86)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/metrics.go, line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLCwG907Qf5mbRKNDg:-LaYnsWE72ntfPlb1PLZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L98)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> Number of deadlocks detected by the txn wait queue.
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/txnqueue.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaIHZnY8QrvTLc0lM-V:-LaYnvpS7FD32h0WdQuV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L473)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Please use the same pattern from

https://github.com/cockroachdb/cockroach/blob/38e796199562090d2df956ad643e71b25ee25915/pkg/storage/replica_write.go#L194-L223

which is allocation optimized (and prevents deadlocks).

@andy-kimball if you enter the queue, a lot more work is already being done than this txn, so it shouldn't be on the critical perf path.
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/txnqueue.go, line 481 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLEpRf6d5fuL2j-wzq:-LaYnxpq9MYj7ka4KBD3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L481)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Instead of calling `RecordValue` for this in many different places, `defer` a function here that does it.
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/txnqueue.go, line 485 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLDm5-0gE8-9dgJhjD:-LaYnzsT1n4FKNgWyGXK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L485)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Please also use the same `have been waiting` kind of wording and print the pusher and pushee. You also need to `defer` a decrement to the gauge. Basically just do exactly what that other code I refer to above does.
</blockquote></details>

Done.
___
*[pkg/storage/txnwait/txnqueue.go, line 694 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLFAucENMsOvlH1f4O:-LaYo1P5925EwDefjDDZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L694)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ditto about `defer`ing this once instead.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Sorry it took me so long to get back at this. Still mostly commenting on style, but looks like this is getting there. I'll want to give all of the metrics operations another good look to see if I can spot any way in which they could diverge.

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-:-LabJj4z4lOuc5Zt8Kcu:b-g77tf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/store.go, line 4357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab5bSe0yFwRqpdCzY-:-Lab5bSf-sdDJk1_SdY6:b4u4vk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/store.go#L4357)):*
> txnwait

(not txnqueue)
___
*[pkg/storage/txn_wait_queue_test.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab5i231e-MpXEU0D0K:-Lab5i231e-MpXEU0D0L:b-nchl5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L58)):*
> ```Go
> func assertAllGaugesAreZero(tc testContext) {
> 	testutils.SucceedsSoon(tc.TB, func() error {
> 		if !tc.store.GetTxnWaitMetrics().HasAllGaugesZero() {
> ```

`HasAllGaugesZero` should live in this file.
___
*[pkg/storage/txn_wait_queue_test.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab5mHW-zIgc532pqdW:-Lab5mHW-zIgc532pqdX:b-4fc8iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L68)):*
> ```Go
> 	testutils.SucceedsSoon(tc.TB, func() error {
> 		v := tc.store.GetTxnWaitMetrics().PusherWaiting.Value()
> 		if v != expectedCount {
> ```

You can use `assert.Equal(t, expectedCount, v, "unexpected number of pushers")` (here and everywhere else). Not mandatory.
___
*[pkg/storage/txn_wait_queue_test.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab63Qr8Pxa2n-tNAz7:-Lab63Qs8j7p6wZ4skvO:b52zg2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L131)):*
> ```Go
> 		t.Errorf("expected push txn queue is enabled")
> 	}
> 	assertAllGaugesAreZero(tc)
> ```

It doesn't seem that this needs an internal `SucceedsSoon`.
___
*[pkg/storage/txn_wait_queue_test.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab65ly9nCbu6vCItDE:-Lab65ly9nCbu6vCItDF:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L137)):*
> ```Go
> 		t.Fatalf("expected pendingTxn to be in txns map after enqueue")
> 	}
> 	assertPusheeCount(tc, 1)
> ```

Ditto
___
*[pkg/storage/txn_wait_queue_test.go, line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab66RR5YHd62r7WwGt:-Lab66RR5YHd62r7WwGu:bb96zp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L157)):*
> ```Go
> 			return errors.Errorf("expected GetDependents %+v; got %+v", expDeps, deps)
> 		}
> 		assertPusheeCount(tc, 1)
> ```

Neither do these two, because it's in a `SucceedsSoon` already (but you need to return an error from `assertPusherCount`)

I think you would be even better off without all these helpers and just writing `m := tc.store.TxnWaitMetrics()` and then in the code

```go
assert.Equal(t, exp, m.PusherWaiting.Count())
```

or if you're in a `SucceedsSoon`

```go
if act := m.PusherWaiting.Count(); act != exp {
  return errors.Errorf("%d pushers, but want %d", act, exp)
}
```
___
*[pkg/storage/txn_wait_queue_test.go, line 221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LabIEFfAtLD1AL-YnxK:-LabIEFfAtLD1AL-YnxL:b32a99j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L221)):*
> ```Go
> 	assertAllGaugesAreZero(tc)
> 	q.Enqueue(txn)
> 	assertPusheeCount(tc, 1)
> ```

this would then read `assert.Equal(t, 1, m.PusheeCount.Value())`
___
*[pkg/storage/txnwait/metrics.go, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LabISwN7FT5da5P1fPL:-LabISwOEfR-DHjuBYvN:b-h2ofs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txnwait/metrics.go#L49)):*
> ```Go
> 				Name:        "txnwaitqueue.pushee.waiting",
> 				Help:        "Number of pushees on the txn wait queue",
> 				Measurement: "PusheeWaiting",
> ```

Is that what measurement strings look like? I think this is what gets shown on the graph axis. Maybe "Waiting Pushees"? Ditto for the other metrics.
___
*[pkg/storage/txnwait/txnqueue.go, line 485 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLDm5-0gE8-9dgJhjD:-LabJMtv6n0YVAluiMzZ:b-iew654) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L485)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: `pusher %s: have been waiting ...` rolls off the tongue better. I'd also unify the two cases (using the empty uuid when no pusher) simply because we'll easily see what's going on from the log message and also in practice there will always be a txn (we use nontransactional reqs for barely anything).


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/store.go, line 4357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab5bSe0yFwRqpdCzY-:-LagvJVEAaxxWHR2JAai:brythjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/store.go#L4357)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> txnwait

(not txnqueue)
</blockquote></details>

renamed.
___
*[pkg/storage/txn_wait_queue_test.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab5i231e-MpXEU0D0K:-Lagv_z15jqulwYEUauE:b-ne7jz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L58)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`HasAllGaugesZero` should live in this file.
</blockquote></details>

moved.
___
*[pkg/storage/txn_wait_queue_test.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab5mHW-zIgc532pqdW:-Lagvtxw7bPcRX-Z78QW:bduy7y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L68)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You can use `assert.Equal(t, expectedCount, v, "unexpected number of pushers")` (here and everywhere else). Not mandatory.
</blockquote></details>

I changed it.
___
*[pkg/storage/txn_wait_queue_test.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab63Qr8Pxa2n-tNAz7:-LagvyEEDbzCOj6cNj21:b-tra26k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L131)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It doesn't seem that this needs an internal `SucceedsSoon`.
</blockquote></details>

I agree that it doesn't need. I changed to be explicit. But I do think that this is worse than the way it was before. What I want to test is that gauges are set in a certain way at the point of the test. Whether they are being set synchronously in a direct call or in a go routine or the modifications are deferred - isn't important. A switch from one to the other shouldn't require me to modify tests. As is right now - it will require.
___
*[pkg/storage/txn_wait_queue_test.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab65ly9nCbu6vCItDE:-LagxFQ39nK6hjkhpaJI:b3qgw97) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L137)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

fixed too.
___
*[pkg/storage/txn_wait_queue_test.go, line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab66RR5YHd62r7WwGt:-LagxHiY8lRGP3AXuCU7:b-7ngf2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L157)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Neither do these two, because it's in a `SucceedsSoon` already (but you need to return an error from `assertPusherCount`)

I think you would be even better off without all these helpers and just writing `m := tc.store.TxnWaitMetrics()` and then in the code

```go
assert.Equal(t, exp, m.PusherWaiting.Count())
```

or if you're in a `SucceedsSoon`

```go
if act := m.PusherWaiting.Count(); act != exp {
  return errors.Errorf("%d pushers, but want %d", act, exp)
}
```
</blockquote></details>

changed too. 
___
*[pkg/storage/txn_wait_queue_test.go, line 221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LabIEFfAtLD1AL-YnxK:-LagxatS9dWoxMDy62Q7:b-b5c35t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L221)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

this would then read `assert.Equal(t, 1, m.PusheeCount.Value())`
</blockquote></details>

changed.


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txnwait/metrics.go, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LabISwN7FT5da5P1fPL:-Lagxh_p1wX3pg3sjOkn:b-lfnh4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txnwait/metrics.go#L49)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is that what measurement strings look like? I think this is what gets shown on the graph axis. Maybe "Waiting Pushees"? Ditto for the other metrics.
</blockquote></details>

Good catch. I didn't see that this is used as labels on the graphs. It makes more sense that way you suggest.


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txnwait/txnqueue.go, line 485 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaLDm5-0gE8-9dgJhjD:-LagyAj9ChCYvfjBGu-p:bbws5a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/txnqueue.go#L485)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: `pusher %s: have been waiting ...` rolls off the tongue better. I'd also unify the two cases (using the empty uuid when no pusher) simply because we'll easily see what's going on from the log message and also in practice there will always be a txn (we use nontransactional reqs for barely anything).
</blockquote></details>

Modified the way you are suggesting. Agreed that it is better.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

You're going to want to add a category to the release note.

Something like

> Release note (admin ui change): Added txn wait queue related metrics for debugging (available only via the custom chart page).

> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-:-LaoLOEv2rxbTCqWx9Iq:bb3r4ex)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, and @nvanbenschoten)

___
*[pkg/storage/txn_wait_queue_test.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab63Qr8Pxa2n-tNAz7:-LaoFj3S4oSYBJYXeiS4:bicmbhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L131)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

I agree that it doesn't need. I changed to be explicit. But I do think that this is worse than the way it was before. What I want to test is that gauges are set in a certain way at the point of the test. Whether they are being set synchronously in a direct call or in a go routine or the modifications are deferred - isn't important. A switch from one to the other shouldn't require me to modify tests. As is right now - it will require.
</blockquote></details>

By that same reasoning, should everything we assert through all of the unit tests in the code base tolerate asynchronicity? I think it is better to assume no asynchronicity by default and to adapt only when it is found to be necessary. We should also generally avoid `SucceedsSoon` and by that token, not use it unless actually necessary. 
___
*[pkg/storage/txn_wait_queue_test.go, line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaoIBwb35oaIYAT7Ltr:-LaoIBwc08flEe_t_E-_:bvv8u9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf87e96fddb71c7fa31ca0b467d3b6c7d78c048/pkg/storage/txn_wait_queue_test.go#L57)):*
> ```Go
> }
> 
> func hasAllGaugesZero(tc testContext) bool {
> ```

When this test fails, it'll be unclear which gauge exactly triggered it. I'd do this as

```go
func hasAllGaugesZero(m *txnwait.Metrics) error {
   if act := m.PusheeWaiting.Value(); act != 0 { return errors.Errorf("PusheeWaiting=%d", act) }
  ...
   return nil
}
```

All of this is unfortunately a bit verbose, and can be done better, but let's keep it basic for now.
___
*[pkg/storage/txnwait/metrics.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaMdfb9-hUXMGIz_mOh:-LaoH7g2Arl2oehnJNll:bkyvqpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L56)):*
`QueryTxn` loads the pusher's own transaction (and I guess the pushee's too) on some timer to make sure they eventually get an up to date view. It also returns the txns waiting on the query-ee which I guess is used for deadlock detection. The txnwaitqueue also lets QueryTxn wait until the txn record updates relative to what the querier provided. I can't say that I understand it any better than that. In light of that maybe

> Number of transaction status queries waiting for an updated transaction record
___
*[pkg/storage/txnwait/metrics.go, line 77 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaoJMCsEi-o6TjfiDwB:-LaoJMCsEi-o6TjfiDwC:b2k7jc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf87e96fddb71c7fa31ca0b467d3b6c7d78c048/pkg/storage/txnwait/metrics.go#L77)):*
> ```Go
> 			metric.Metadata{
> 				Name:        "txnwaitqueue.pusher.wait_time",
> 				Help:        "Histogram of durations spent in queue by pushers.",
> ```

Figure out what the convention is (trailing dot or not?) and use the same throughout. I think it's no trailing dot based on what I see in `storage/metrics.go`.
___
*[pkg/storage/txnwait/txnqueue.go, line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaoKJPt0up7uACVskPq:-LaoKJPu850lCzCvylly:bjh4wc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf87e96fddb71c7fa31ca0b467d3b6c7d78c048/pkg/storage/txnwait/txnqueue.go#L501)):*
> ```Go
> 				timeutil.Since(tBegin).Seconds(),
> 				req.PusheeTxn.ID.Short(),
> 			)
> ```

In other places, we print a corresponding message that gives the actual wait time. I'd do something similar here by `defer`'ing a closure that prints the time again when it fires.

See

https://github.com/cockroachdb/cockroach/blob/cf449c6cb910efe646e118d3d5c9c3561765cd24/pkg/storage/replica_write.go#L216-L225


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
updated the release note category.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-:-LagyNXn6g2ijtt8Ljf7:bi80b4k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @andy-kimball, @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txn_wait_queue_test.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-Lab63Qr8Pxa2n-tNAz7:-Laq-ML04UNQmN5TWXwB:b-gyrvub) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e5eed79b20c4a63ff95af0516d1fe58ea4fb33/pkg/storage/txn_wait_queue_test.go#L131)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

By that same reasoning, should everything we assert through all of the unit tests in the code base tolerate asynchronicity? I think it is better to assume no asynchronicity by default and to adapt only when it is found to be necessary. We should also generally avoid `SucceedsSoon` and by that token, not use it unless actually necessary. 
</blockquote></details>

No, I don't think everything should be tested in a way that tolerates asynchronicity. I could imagine many cases where deferring the result will be a bug so the test shouldn't allow for immediate failure and success after a wait and retry. However for the metrics case above this issue doesn't exist and all metrics reporting should ideally be deferred to not slow down the main code. 
___
*[pkg/storage/txn_wait_queue_test.go, line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaoIBwb35oaIYAT7Ltr:-Laq-Hc35egRSNXVMeR-:bffveur) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf87e96fddb71c7fa31ca0b467d3b6c7d78c048/pkg/storage/txn_wait_queue_test.go#L57)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

When this test fails, it'll be unclear which gauge exactly triggered it. I'd do this as

```go
func hasAllGaugesZero(m *txnwait.Metrics) error {
   if act := m.PusheeWaiting.Value(); act != 0 { return errors.Errorf("PusheeWaiting=%d", act) }
  ...
   return nil
}
```

All of this is unfortunately a bit verbose, and can be done better, but let's keep it basic for now.
</blockquote></details>

Refactored it. I agree that this is better.
___
*[pkg/storage/txnwait/metrics.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaMdfb9-hUXMGIz_mOh:-Laq1-sq0w717erglqrT:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0527fbc3eb397031c8f60010c8211c2ecbfd3d0/pkg/storage/txnwait/metrics.go#L56)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`QueryTxn` loads the pusher's own transaction (and I guess the pushee's too) on some timer to make sure they eventually get an up to date view. It also returns the txns waiting on the query-ee which I guess is used for deadlock detection. The txnwaitqueue also lets QueryTxn wait until the txn record updates relative to what the querier provided. I can't say that I understand it any better than that. In light of that maybe

> Number of transaction status queries waiting for an updated transaction record
</blockquote></details>

Changed.
___
*[pkg/storage/txnwait/metrics.go, line 77 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaoJMCsEi-o6TjfiDwB:-Laq11fX2Npbawc1maAa:bpwn71a) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf87e96fddb71c7fa31ca0b467d3b6c7d78c048/pkg/storage/txnwait/metrics.go#L77)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Figure out what the convention is (trailing dot or not?) and use the same throughout. I think it's no trailing dot based on what I see in `storage/metrics.go`.
</blockquote></details>

No dot seems to be the default. I fixed the few offenders. 
___
*[pkg/storage/txnwait/txnqueue.go, line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaoKJPt0up7uACVskPq:-Laq194R3q5RjCxhdH8k:b-a2ujfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf87e96fddb71c7fa31ca0b467d3b6c7d78c048/pkg/storage/txnwait/txnqueue.go#L501)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

In other places, we print a corresponding message that gives the actual wait time. I'd do something similar here by `defer`'ing a closure that prints the time again when it fires.

See

https://github.com/cockroachdb/cockroach/blob/cf449c6cb910efe646e118d3d5c9c3561765cd24/pkg/storage/replica_write.go#L216-L225
</blockquote></details>

Changed.


<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
Thanks. Almost ready!

> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-:-LaqDF-wDLTTo4iOUDDd:b-bd8d4g)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @andy-kimball, @bdarnell, @darinpp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txnwait/txnqueue.go, line 500 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaqD4XD9cAHzgWcKk4G:-LaqD4XD9cAHzgWcKk4H:beds54j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cebefdc4bd243658ace18ffd5c515f2ee3fb3bb/pkg/storage/txnwait/txnqueue.go#L500)):*
> ```Go
> 			defer func() {
> 				metrics.PusherSlow.Dec(1)
> 				log.Warningf(ctx, "pusher %s finished waiting after %.2fs for %s pushee",
> ```

This message is in addition to the one after one minute. Also please update the wording for both to

`pusher %s: .... for pushee %s`.


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @andy-kimball, @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/storage/txnwait/txnqueue.go, line 500 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35921#-LaqD4XD9cAHzgWcKk4G:-LarDrmcE-y4VtkZFvA9:b90y0z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cebefdc4bd243658ace18ffd5c515f2ee3fb3bb/pkg/storage/txnwait/txnqueue.go#L500)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This message is in addition to the one after one minute. Also please update the wording for both to

`pusher %s: .... for pushee %s`.
</blockquote></details>

Ah..makes sense to log both. Updated.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35921)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @andy-kimball, @bdarnell, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r+

Thanks @darinpp! After this has merged, could you backport it to `release-19.1`? You can do so by running `backport 35921` after having installed `backport` via `go get github.com/benesch/backport`.
